### PR TITLE
Add `@feature` directive to conditionally add a field based on a feature state (using Laravel Pennant)

### PR DESCRIPTION
**Pull request by remipelhate**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

**Changes**

This PR introduces a new `@feature` directive, which can be used to conditionally add a field to the schema based on the state of a feature flag:

```graphql
query {
  fieldWhenFeatureInactive: String! 
    @feature(name: "new-api", when: INACTIVE)

  fieldWhenFeatureActive: String!
    @feature(name: "new-api")
}
```

The `fieldWhenFeatureInactive` will only be available if the feature called `new-api` is inactive, whereas fieldWhenFeatureActive will only be available when the feature is active.

I've been using this in a codebase for a client, and thought it might be useful in other projects as well. Let me know if you guys like this ;)

**Breaking changes**

There are no breaking changes.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -51,6 +51,7 @@
         "laravel/framework": "^9 || ^10",
         "laravel/legacy-factories": "^1.1.1",
         "laravel/lumen-framework": "^9 || ^10 || dev-master",
+        "laravel/pennant": "^1.0",
```
Comment (spawnia) [inline composer.json:54]:
```suggestion
        "laravel/pennant": "^1",
```

File: src/Pennant/FeatureDirective.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        private FeatureManager $features,
+    ) {
+        parent::__construct();
+    }
+
+    public static function definition(): string
+    {
+        return /** @lang GraphQL */ <<<'GRAPHQL'
+"""
+Conditionally add the annotated element to schema depending of the state of a feature using Laravel Pennant.
```
Comment (spawnia) [inline src/Pennant/FeatureDirective.php:21]:
```suggestion
Include the annotated element in the schema depending on a Laravel Pennant feature.
```

File: src/Pennant/FeatureDirective.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+"""
+directive @feature(
+    """
+    The name of the feature to be checked (can be a string or class name).
+    """
+    name: String
```
Comment (spawnia) [inline src/Pennant/FeatureDirective.php:27]:
```suggestion
    name: String!
```

Seems necessary

File: src/Pennant/FeatureDirective.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    
+    """
+    Specify what the state of the feature should be for the field to be applied.
+    """
+    when: FeatureState = ACTIVE
```
Comment (spawnia) [inline src/Pennant/FeatureDirective.php:32]:
```suggestion
    when: FeatureState! = ACTIVE
```

Optional, but non-null

File: src/Pennant/FeatureDirective.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    protected function shouldHide(): bool
+    {
+        $feature = $this->directiveArgValue('name');
+        $requiredFeatureState = $this->directiveArgValue('when', 'ACTIVE');
+
+        return match ($requiredFeatureState) {
+            'ACTIVE' => $this->features->inactive($feature),
+            'INACTIVE' => $this->features->active($feature),
+            default => throw new DefinitionException("Found invalid feature state: {$requiredFeatureState}"),
```
Comment (spawnia) [inline src/Pennant/FeatureDirective.php:60]:
```suggestion
            default => throw new DefinitionException("Expected FeatureState `ACTIVE` or `INACTIVE` for argument `when` of {$this->name()} on {$this->nodeName()}, got `{$requiredFeatureState}`."),
```

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        @feature(name: "foo", when: INACTIVE)
+        @mock
+    fieldWhenFeatureActive: String!
+        @feature(name: "foo", when: ACTIVE)
+        @mock
+    fieldWhenFeatureActiveByDefault: String!
+        @feature(name: "foo")
+        @mock 
+}
+GRAPHQL;
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:31]: Please define a separate schema in each test case, experience has shown that having a single test schema per class gets unwieldy.

**Review by spawnia — CHANGES_REQUESTED** [review]
Nicely done. I am fine with adding this feature. Please add a section to the docs.

**Comment by remipelhate** [comment]
Didn't expect such a quick response👌🏻 Thanks for your feedback! 

Where in the docs would you like me to add a section? I was looking at how the `@search` directive is implemented (as it depends on Laravel Scout) and I can't seem to find it in the docs outside of the API reference (which I've added).

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        $this->schema = /* @lang GraphQL */ <<<'GRAPHQL'
+            type Query {
+                fieldWhenActive: String!
+                    @feature(name: "new-api", when: ACTIVE)
+                    @mock
+            }
+            GRAPHQL;
+
+        $response = $this->graphQL(/* @lang GraphQL */ <<<'GRAPHQL'
+            query {
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:27]:
```suggestion
            {
```

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        $this->schema = /* @lang GraphQL */ <<<'GRAPHQL'
+            type Query {
+                fieldWhenActive: String!
+                    @feature(name: "new-api")
+                    @mock
+            }
+            GRAPHQL;
+
+        $response = $this->graphQL(/* @lang GraphQL */ <<<'GRAPHQL'
+            query {
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:47]:
```suggestion
            {
```

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        $this->schema = /* @lang GraphQL */ <<<'GRAPHQL'
+            type Query {
+                fieldWhenInactive: String!
+                    @feature(name: "new-api", when: INACTIVE)
+                    @mock
+            }
+            GRAPHQL;
+
+        $response = $this->graphQL(/* @lang GraphQL */ <<<'GRAPHQL'
+            query {
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:68]:
```suggestion
            {
```

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        $this->schema = /* @lang GraphQL */ <<<'GRAPHQL'
+            type Query {
+                fieldWhenActive: String!
+                    @feature(name: "new-api", when: ACTIVE)
+                    @mock
+            }
+            GRAPHQL;
+
+        $response = $this->graphQL(/* @lang GraphQL */ <<<'GRAPHQL'
+            query {
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:90]:
```suggestion
            {
```

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        $this->schema = /* @lang GraphQL */ <<<'GRAPHQL'
+            type Query {
+                fieldWhenActive: String!
+                    @feature(name: "new-api")
+                    @mock
+            }
+            GRAPHQL;
+
+        $response = $this->graphQL(/* @lang GraphQL */ <<<'GRAPHQL'
+            query {
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:117]:
```suggestion
            {
```

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        $this->schema = /* @lang GraphQL */ <<<'GRAPHQL'
+            type Query {
+                fieldWhenInactive: String!
+                    @feature(name: "new-api", when: INACTIVE)
+                    @mock
+            }
+            GRAPHQL;
+
+        $response = $this->graphQL(/* @lang GraphQL */ <<<'GRAPHQL'
+            query {
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:143]:
```suggestion
            {
```

File: tests/Integration/Pennant/FeatureDirectiveTest.php [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+                fieldWhenActive
+            }
+            GRAPHQL,
+        );
+
+        $response->assertGraphQLErrorFree();
+        $response->assertJson([
+            'data' => [
+                'fieldWhenActive' => 'active',
```
Comment (spawnia) [inline tests/Integration/Pennant/FeatureDirectiveTest.php:99]: Extract `'active'` to a variable and use it above to clarify the connection between the test data setup and the assertion

**Comment by spawnia** [comment]
> Where in the docs would you like me to add a section? I was looking at how the `@search` directive is implemented (as it depends on Laravel Scout) and I can't seem to find it in the docs outside of the API reference (which I've added).

https://github.com/nuwave/lighthouse/blob/master/docs/master/digging-deeper/feature-toggles.md

**Comment by remipelhate** [comment]
@spawnia Feedback processed. Let me know if anything is unclear.

File: docs/master/digging-deeper/feature-toggles.md [hunk 14]
```diff
@@ -25,9 +25,46 @@ type Query {
...
+Another example would be to only include a field when the feature is inactive:
+
+```graphql
+type Query {
+  deprecatedField: String! @feature(name: "new-api", when: "INACTIVE")
+}
+```
+
+When using [class based features](https://laravel.com/docs/pennant#class-based-features), 
+the classname should be used as value for the `name` argument:
```
Comment (spawnia) [inline docs/master/digging-deeper/feature-toggles.md:52]:
```suggestion
the fully qualified class name must be used as the value for the `name` argument:
```

File: docs/master/digging-deeper/feature-toggles.md [hunk 15]
```diff
@@ -25,9 +25,46 @@ type Query {
...
+}
+```
+
+In this case, `experimentalField` will only be included when the `new-api` feature is active. 
+
+Another example would be to only include a field when the feature is inactive:
+
+```graphql
+type Query {
+  deprecatedField: String! @feature(name: "new-api", when: "INACTIVE")
```
Comment (spawnia) [inline docs/master/digging-deeper/feature-toggles.md:47]:
```suggestion
  deprecatedField: String! @feature(name: "new-api", when: INACTIVE)
```

If not already the case, tests and other examples should also use this as a proper enum.

**Review by spawnia — COMMENTED** [review]
Just some minor tweaks, otherwise good to go.

File: docs/master/digging-deeper/feature-toggles.md (hunk 15 above)
Comment (remipelhate) [reply docs/master/digging-deeper/feature-toggles.md:47]: I double-checked, but this was the only typo. Good catch!

**Comment by spawnia** [comment]
Thanks, released with https://github.com/nuwave/lighthouse/releases/tag/v6.18.0.

**Pull request merged**